### PR TITLE
OHRM5X-1272: Create Language packs

### PR DESCRIPTION
**Pull request by SharukaNP**

### CODE REVIEW

File: installer/Migration/V5_0_0/TranslationHelper.php [hunk 1]
```diff
@@ -0,0 +1,126 @@
+<?php
+
```
Comment (RajithaKumara) [inline installer/Migration/V5_0_0/TranslationHelper.php:2]: copyright header. setup IDE

File: installer/Migration/V5_0_0/TranslationHelper.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * @param string $langCode
+     * @return int
+     * @throws Exception
+     */
+    private function getLanguageId(string $langCode): int
+    {
+        $searchQuery = $this->createQueryBuilder();
+        $searchQuery->select('language.id')->from('ohrm_i18n_language', 'language')->where('language.code = :langCode')->setParameter('langCode', $langCode);
```
Comment (RajithaKumara) [inline installer/Migration/V5_0_0/TranslationHelper.php:97]: format into multiple lines

File: installer/upgrader/Migrations/V5/TranslationGenerateTool.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+use OrangeHRM\Core\Traits\ORM\EntityManagerHelperTrait;
+use Symfony\Component\Yaml\Yaml;
+
+class TranslationGenerateTool
+{
+    use EntityManagerHelperTrait;
+
+    public function generateTranslations()
+    {
+        $langCodes = ['bg_BG', 'da_DK', 'de', 'en_US', 'es', 'es_AR', 'es_BZ', 'es_CR', 'es_ES', 'fr', 'fr_FR', 'id_ID', 'ja_JP', 'nl', 'om_ET', 'th_TH', 'vi_VN', 'zh_Hans_CN', 'zh_Hant_TW', 'zz_ZZ'];  //add the xml files inside installer/upgrader/Migrations/V5/translations/messages folder
```
Comment (RajithaKumara) [inline installer/upgrader/Migrations/V5/TranslationGenerateTool.php:32]:
```suggestion
        $langCodes = ['bg_BG', 'da_DK', 'de', 'en_US', 'es', 'es_AR', 'es_BZ', 'es_CR', 'es_ES', 'fr', 'fr_FR', 'id_ID', 'ja_JP', 'nl', 'om_ET', 'th_TH', 'vi_VN', 'zh_Hans_CN', 'zh_Hant_TW'];  //add the xml files inside installer/upgrader/Migrations/V5/translations/messages folder
```

File: installer/upgrader/runTranslations.php [hunk 4]
```diff
@@ -19,16 +19,16 @@
...
 
 ServiceContainer::getContainer()->register(Services::DOCTRINE)
     ->setFactory([Doctrine::class, 'getEntityManager']);
 
-$translate = new TranslationTool();
-$modules = ['admin','general','pim','leave','time','attendance','maintenance','help','auth'];
-foreach ($modules as $module){
-    $translate->up($module);
+$translate = new TranslationHelper();
+$langCodes = ['bg_BG', 'da_DK', 'de', 'en_US', 'es', 'es_AR', 'es_BZ', 'es_CR', 'es_ES', 'fr', 'fr_FR', 'id_ID', 'ja_JP', 'nl', 'om_ET', 'th_TH', 'vi_VN', 'zh_Hans_CN', 'zh_Hant_TW', 'zz_ZZ'];
```
Comment (RajithaKumara) [inline installer/upgrader/runTranslations.php:31]:
```suggestion
$langCodes = ['bg_BG', 'da_DK', 'de', 'en_US', 'es', 'es_AR', 'es_BZ', 'es_CR', 'es_ES', 'fr', 'fr_FR', 'id_ID', 'ja_JP', 'nl', 'om_ET', 'th_TH', 'vi_VN', 'zh_Hans_CN', 'zh_Hant_TW'];
```

File: installer/upgrader/generateTranslations.php [hunk 5]
```diff
@@ -19,13 +19,13 @@
 
 use OrangeHRM\Framework\ServiceContainer;
 use OrangeHRM\Framework\Services;
+use OrangeHRM\Installer\upgrader\Migrations\V5\TranslationGenerateTool;
```
Comment (RajithaKumara) [inline installer/upgrader/generateTranslations.php:22]:
```suggestion
use OrangeHRM\Tools\Migrations\V5\TranslationGenerateTool;
```

File: installer/upgrader/Migrations/V5/TranslationTestTool.php [hunk 6]
```diff
@@ -19,12 +19,11 @@
 
 namespace OrangeHRM\Tools\Migrations\V5;
 
-use Doctrine\DBAL\Connection;
 use Doctrine\DBAL\Exception;
 use Doctrine\DBAL\Query\QueryBuilder;
 use OrangeHRM\Core\Traits\ORM\EntityManagerHelperTrait;
 use OrangeHRM\Installer\Migration\V5_0_0\LangStringHelper;
-use OrangeHRM\Installer\Util\V1\Dto\TransUnit;
+use OrangeHRM\Installer\upgrader\Migrations\V5\TranslationUnit;
```
Comment (RajithaKumara) [inline installer/upgrader/Migrations/V5/TranslationTestTool.php:26]:
```suggestion
use OrangeHRM\Tools\Migrations\V5\TranslationUnit;
```

File: installer/upgrader/Migrations/V5/TranslationGenerateTool.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ * OrangeHRM is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with this program;
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Installer\upgrader\Migrations\V5;
```
Comment (RajithaKumara) [inline installer/upgrader/Migrations/V5/TranslationGenerateTool.php:20]:
```suggestion
namespace OrangeHRM\Tools\Migrations\V5;
```

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Comment by RajithaKumara** [comment]
Need to move all classes away from `OrangeHRM\Tools\Migrations\V5` namespace. and delete `installer/upgarder` folder. should move those tools somewhere else

**Comment by SharukaNP** [comment]
> Need to move all classes away from `OrangeHRM\Tools\Migrations\V5` namespace. and delete `installer/upgarder` folder. should move those tools somewhere else



> Need to move all classes away from `OrangeHRM\Tools\Migrations\V5` namespace. and delete `installer/upgarder` folder. should move those tools somewhere else

I'll put into devtools/i18n ?

File: devTools/core/src/Command/AddTestTranslationCommand.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
+<?php
+
```
Comment (RajithaKumara) [inline devTools/core/src/Command/AddTestTranslationCommand.php:2]: copyright header

File: devTools/core/src/Command/AddTestTranslationCommand.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    private TranslationTestTool $translationTestTool;
+
+    /**
+     * @inheritDoc
+     */
+    protected function configure(): void
+    {
+        // ...
+    }
```
Comment (RajithaKumara) [inline devTools/core/src/Command/AddTestTranslationCommand.php:23]:
```suggestion
```

File: devTools/core/src/Command/TranslationCreate.php [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+    private TranslationHelper $translationHelper;
+
+    /**
+     * @inheritDoc
+     */
+    protected function configure(): void
+    {
+        // ...
+    }
```
Comment (RajithaKumara) [inline devTools/core/src/Command/TranslationCreate.php:43]:
```suggestion
```

File: devTools/core/src/Command/Util/TranslationTestTool.php [hunk 11]
```diff
@@ -56,19 +54,19 @@ private function addTranslations(string $language, string $groupName): void
...
     /**
      * @param string $groupName
-     * @param TransUnit $source
+     * @param TranslationUnit $source
      * @param string $language
      * @return void
      * @throws Exception
      */
-    private function saveTranslationRecord(string $groupName, TransUnit $source, string $language): void
+    private function saveTranslationRecord(string $groupName, TranslationUnit $source, string $language): void
```
Comment (RajithaKumara) [inline devTools/core/src/Command/Util/TranslationTestTool.php:69]: using `OrangeHRM\Installer\Util\V1\Dto\TransUnit` here is fine, no need to create another class `TranslationUnit`

**Review by RajithaKumara — CHANGES_REQUESTED** [review]
I think better add readme how to use console commands with few examples. also where developer should place xml

File: devTools/core/src/Command/Util/TranslationTestTool.php (hunk 11 above)
Comment (SharukaNP) [reply devTools/core/src/Command/Util/TranslationTestTool.php:69]: The classes are a bit different, In TranslationUnit I need the to have the source as one of its attributes as the target is generated using the source.

**Comment by SharukaNP** [comment]
Added readme Testing on both scenarios work

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**